pkg/cmd/repo/sync/sync.go

```diff
@@ -0,0 +1,322 @@
+		fmt.Fprintf(opts.IO.Out, "%s %s", cs.SuccessIcon(), success)
+	}
+
+	return nil
+}
+
+func syncRemoteRepo(opts *SyncOptions) error {
+	httpClient, err := opts.HttpClient()
+	if err != nil {
+		return err
+	}
+	apiClient := api.NewClientFromHTTP(httpClient)
+
+	var destRepo, srcRepo ghrepo.Interface
+
+	destRepo, err = ghrepo.FromFullName(opts.DestArg)
+	if err != nil {
+		return err
+	}
+
+	if opts.SrcArg == "" {
+		opts.IO.StartProgressIndicator()
+		srcRepo, err = api.RepoParent(apiClient, destRepo)
+		opts.IO.StopProgressIndicator()
+		if err != nil {
+			return err
+		}
+		if srcRepo == nil {
+			return fmt.Errorf("can't determine source repo for %s because repo is not fork", ghrepo.FullName(destRepo))
+		}
+	} else {
+		srcRepo, err = ghrepo.FromFullName(opts.SrcArg)
+		if err != nil {
+			return err
+		}
+	}
+
+	if destRepo.RepoHost() != srcRepo.RepoHost() {
+		return fmt.Errorf("can't sync repos from different hosts")
+	}
+
+	if opts.Branch == "" {
+		opts.IO.StartProgressIndicator()
+		opts.Branch, err = api.RepoDefaultBranch(apiClient, srcRepo)
+		opts.IO.StopProgressIndicator()
+		if err != nil {
+			return err
+		}
+	}
+
+	opts.IO.StartProgressIndicator()
+	err = executeRemoteRepoSync(apiClient, destRepo, srcRepo, opts)
+	opts.IO.StopProgressIndicator()
+	if err != nil {
+		if errors.Is(err, divergingError) {
+			return fmt.Errorf("can't sync because there are diverging changes, you can use `--force` to overwrite the changes")
+		}
+		return err
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		srcStr := fmt.Sprintf("%s:%s", ghrepo.FullName(srcRepo), opts.Branch)
+		destStr := fmt.Sprintf("%s:%s", ghrepo.FullName(destRepo), opts.Branch)
+		success := fmt.Sprintf("Synced %s from %s\n", cs.Bold(destStr), cs.Bold(srcStr))
+		fmt.Fprintf(opts.IO.Out, "%s %s", cs.SuccessIcon(), success)
+	}
+
+	return nil
+}
+
+var divergingError = errors.New("diverging changes")
+var mismatchRemotesError = errors.New("branch remote does not match specified source")
+
+func executeLocalRepoSync(srcRepo ghrepo.Interface, remote string, opts *SyncOptions) error {
+	git := opts.Git
+	branch := opts.Branch
+
+	err := git.Fetch([]string{remote, fmt.Sprintf("+refs/heads/%s", branch)})
+	if err != nil {
+		return err
+	}
+
+	hasLocalBranch := git.HasLocalBranch([]string{branch})
+	if hasLocalBranch {
+		branchRemote, err := git.BranchRemote(branch)
+		if branchRemote != remote {
+			return mismatchRemotesError
+		}
+
+		fastForward, err := git.IsAncestor([]string{branch, fmt.Sprintf("%s/%s", remote, branch)})
+		if err != nil {
+			return err
+		}
+
+		if !fastForward && !opts.Force {
+			return divergingError
+		}
+	}
+
+	startBranch, err := git.CurrentBranch()
+	if err != nil {
+		return err
+	}
+	if startBranch != branch {
+		if hasLocalBranch {
+			err = git.Checkout([]string{branch})
+			if err != nil {
+				return err
+			}
+		} else {
+			err = git.Checkout([]string{"--track", fmt.Sprintf("%s/%s", remote, branch)})
```

Comment:
This should ensure that if there are multiple remotes with the same branch name we will checkout the correct one.

Reply:
Shouldn't the invocation be: `git checkout <branch> --track <remote>/<branch>`? It looks like the local branch argument is missing

Reply:
According to the docs (https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt---track) the name to use for the local branch will be automatically determined.